core/arch/riscv/kernel/entry.S

```diff
@@ -231,13 +234,27 @@ END_FUNC reset_primary
 LOCAL_FUNC reset_secondary , : , .identity_map
 UNWIND(	.cantunwind)
 	wait_primary
+
+#ifdef CFG_RISCV_S_MODE
+.option push
+.option norelax
+	la	gp, __global_pointer$
+.option pop
+	csrw	CSR_XSCRATCH, a0
+#endif
+
 	csrw	CSR_SATP, zero
 	set_sp
 	set_tp
 	set_satp
 	cpu_is_ready
 
 	jal	boot_init_secondary
+#ifdef CFG_RISCV_S_MODE
+	/* Return to untrusted domain */
+	li	a0, TEEABI_OPTEED_RETURN_ON_DONE
+	j	thread_return_to_ree
```

Comment:
@gagachang for this I guess we need to add an extra flag, to decide to use a Security Monitor based solution, or not (could be messaging based as we use today without the need for a SM).
I also suggest to use `thread_return_to_udomain` to keep same wording accross the work.

Reply:
> @gagachang for this I guess we need to add an extra flag, to decide to use a Security Monitor based solution, or not (could be messaging based as we use today without the need for a SM).

It is decided in the implementation of `thread_return_to_ree()`.
Please see my previous commit https://github.com/OP-TEE/optee_os/commit/0fb22936229d51e005aaf159e628e516a60faaa5
If Security Monitor based solution is adopted, `CFG_RISCV_SBI` should be defined and `ecall` is used for communication.
The messaging based solution can be also implemented into `thread_return_to_ree()` by defining another configuration instead of `CFG_RISCV_SBI`.

```C
FUNC thread_return_to_ree , :
	/* Caller should provide arguments in a0~a5 */
#if defined(CFG_RISCV_SBI)
	li	a7, SBI_EXT_TEE		/* extension ID */
	li	a6, 0			/* function ID (unused) */
	ecall
#else
	/* Other protocol */
#endif
	/* ABI to REE should not return */
	panic_at_abi_return
END_FUNC thread_return_to_ree
```

> I also suggest to use `thread_return_to_udomain` to keep same wording accross the work.

OK I agree, "domain" is more precise terminology in RISC-V.
I can rename those `thread_return_to_ree()` to `thread_return_to_udomain()` in another PR.

Reply:
`CFG_RISCV_SBI` is still needed when running OP-TEE OS in S-Mode, regardless how trusted and trusted domain communicate. In your case you are using a specific extension in SBI to context switch between both worlds, but SBI is nedded for other supervisor operations.

Reply:
> `CFG_RISCV_SBI` is still needed when running OP-TEE OS in S-Mode, regardless how trusted and trusted domain communicate. In your case you are using a specific extension in SBI to context switch between both worlds, but SBI is nedded for other supervisor operations.

Yes you are right. SBI has a bunch of extensions, not only for domain communication.
How about `CFG_MSG_BY_SBI` and `CFG_MSG_BY_RPMSG` ? (Assume message based protocol named RPMSG)

Reply:
For ARM, there is a flag `CFG_WITH_ARM_TRUSTED_FW` when we make use of ATF as SM. I propose, for your case, `CFG_RISCV_WITH_SM` that depends on `CFG_RISCV_SBI` maybe ? the addition for messaging based solution we will not be pushed soon. Feel free to enable your SM based code with that flag.

Reply:
> I propose, for your case, `CFG_RISCV_WITH_SM` that depends on `CFG_RISCV_SBI` maybe ?

The [RISC-V CoVE](https://github.com/riscv-non-isa/riscv-ap-tee/tree/main) specification has already defined two kinds of SM, one runs as M-mode and another runs as HS-mode.
Currently we have not supported OP-TEE with hypervisor extension.
How about `CFG_RISCV_WITH_M_MODE_FW` ? It may be more precise.

Reply:
`CFG_RISCV_WITH_M_MODE_FW` is similar to `CFG_RISCV_SBI` an M-Mode FW is needed regardless of the the SM extension. Maybe `CFG_RISCV_WITH_M_MODE_SM` is more precise.

Reply:
> `CFG_RISCV_WITH_M_MODE_FW` is similar to `CFG_RISCV_SBI` an M-Mode FW is needed regardless of the the SM extension. Maybe `CFG_RISCV_WITH_M_MODE_SM` is more precise.

OK, now I know "SM" means the communication is handled by M-mode FW.
So your messaging based solution has M-mode "FW" but not M-mode "SM", right ?
I will add `CFG_RISCV_WITH_M_MODE_SM`.